### ISSUE TITLE
shell script action with administrator privileges

### ISSUE THREAD

**Issue by swiftster** [dustinrue/ControlPlane#289]
I have a few Control Plane actions that require sudo shell commands.  I would like the option to run these with administer privileges (either with or w/out password prompt).  The current workaround is to put it in an applescript "do shell script" wrapper.

**Comment by asfaltboy** [comment]
I have the same requirement, and it took me a while to understand how to use applescript workaround (as suggested by @swiftster), however the action still fails with an error, and does not display a password prompt when executed from ControlPlane. The following should have been a working example, but it isn;'t.

Let's say, you would like to create a script to change DNS servers when arriving to work/home, to the network's gateway as first, followed by google as fallback.

First create an applescript that accepts command line arguments, and sets the DNS popping a password prompt when required.

```applescript
# ~/bin/change_dns.scpt

on run first_dns
	set cmd to "networksetup -setdnsservers Wi-Fi " & first_dns & " 8.8.8.8 8.8.4.4"
	do shell script cmd with administrator privileges
end run
```

Now, since ControlPanel requires the script to be executable with /bin/sh, add a shell script to execute the above script with given arguments.

```bash
# ~/bin/change-dns-via-applescript
osascript ~/bin/change_dns.scpt $@
```

And now, configure the script parameter of your ControlPanel action to:

    ~/bin/change-dns-via-applescript 192.168.2.253

However the script still fails with an error
> 22:37:06.826 -[CPController activateContext:] 	Triggering arrival actions, if any, for 'Home downstairs'
> 22:37:06.829 __56-[ShellScriptAction launchTaskWithLaunchPath:arguments:]_block_invoke 	Failed to execute '/Users/big/bin/change-dns-via-applescript 192.168.1.254' (script terminated with a > non-zero status '127')

It would be great to be able to execute a script with or without a password

**Comment by dustinrue** [comment]
"Value 127 is returned by your shell /bin/bash when any given command within your bash script or on bash command line is not found in any of the paths defined by PATH system environment variable.”

Make sure you have explicitly set your path or you include the full path to all commands being run. This should help.

**Comment by asfaltboy** [comment]
@dustinrue you're right! Setting `/Users/big/bin/change-dns-via-applescript` as the parameter alone works.

I assumed additional arguments are passed to script, but apparently the whole value is evaluated as the path to script. Can we maybe have ShellScript parameter accept arguments? Alternatively, triggering an action could (optionally) pass arguments with information about the activated context?

**Comment by tine2k** [comment]
I managed to get @73's version working (on Sierra with CP 1.6.6). I used a wrapper script that I call from Controlplane. That file `startNfsd.sh` contains the sudo command: `/usr/bin/sudo /sbin/nfsd start`

For some reason ControlPlane fails to call sudo directly.

As far as sudoers go, it is enough to only keep `nfsd` in the sudoers file. The wrapper script can still run unprivileged.

**Comment by dustinrue** [comment]
Not feasible in a sandboxed app. Running shell commands with elevated privileges requires a privileged helper tool (SMJobBless), which adds significant complexity and a separate install step. Out of scope for the rewrite.
